subsys/net/lib/nrf_cloud/coap/src/nrf_cloud_coap.c

```diff
@@ -0,0 +1,489 @@
+	}
+	return err;
+}
+
+int nrf_cloud_coap_location_send(const struct nrf_cloud_gnss_data *gnss)
+{
+	__ASSERT_NO_MSG(gnss != NULL);
+	CHECK_AUTH();
+	int64_t ts = get_ts();
+	size_t len = sizeof(buffer);
+	int err;
+
+	if (gnss->type != NRF_CLOUD_GNSS_TYPE_PVT) {
+		LOG_ERR("Only PVT format is supported");
+		return -ENOTSUP;
+	}
+	err = coap_codec_pvt_encode("GNSS", &gnss->pvt, ts, buffer, &len,
+				    COAP_CONTENT_FORMAT_APP_CBOR);
+	if (err) {
+		LOG_ERR("Unable to encode GNSS PVT data: %d", err);
+		return err;
+	}
+	err = nrf_cloud_coap_post("msg/d2c", NULL, buffer, len,
+				  COAP_CONTENT_FORMAT_APP_CBOR, false, NULL, NULL);
+	if (err) {
+		LOG_ERR("Failed to send POST request: %d", err);
+	}
+	return err;
+}
+
+static int loc_err;
+
+static void get_location_callback(int16_t result_code,
+			 size_t offset, const uint8_t *payload, size_t len,
+			 bool last_block, void *user)
+{
+	LOG_DBG("result_code: %d.%02d, offset:0x%X, len:0x%X, last_block:%d",
+		result_code / 32u, result_code & 0x1f, offset, len, last_block);
+	if (result_code != COAP_RESPONSE_CODE_CONTENT) {
+		loc_err = result_code;
+	} else {
+		loc_err = coap_codec_ground_fix_resp_decode(user, payload, len,
+							    COAP_CONTENT_FORMAT_APP_CBOR);
+	}
+}
+
+int nrf_cloud_coap_location_get(struct nrf_cloud_rest_location_request const *const request,
+				struct nrf_cloud_location_result *const result)
+{
+	__ASSERT_NO_MSG(request != NULL);
+	__ASSERT_NO_MSG((request->cell_info != NULL) || (request->wifi_info != NULL));
+	__ASSERT_NO_MSG(result != NULL);
+	CHECK_AUTH();
+	size_t len = sizeof(buffer);
+	int err;
+
+	err = coap_codec_ground_fix_req_encode(request->cell_info, request->wifi_info, buffer, &len,
+					       COAP_CONTENT_FORMAT_APP_CBOR);
+	if (err) {
+		LOG_ERR("Unable to encode cell pos data: %d", err);
+		return err;
+	}
+	err = nrf_cloud_coap_fetch("loc/ground-fix", NULL, buffer, len,
+				   COAP_CONTENT_FORMAT_APP_CBOR,
+				   COAP_CONTENT_FORMAT_APP_CBOR, true,
+				   get_location_callback, result);
+
+	if (!err && !loc_err) {
+		LOG_DBG("Location: %d, %.12g, %.12g, %d", result->type,
+			result->lat, result->lon, result->unc);
+	} else if (err == -EAGAIN) {
+		LOG_ERR("Timeout waiting for location");
+	} else {
+		LOG_ERR("Error getting location; loc_err:%d, err:%d", loc_err, err);
+		err = loc_err;
+	}
+
+	return err;
+}
+
+static int fota_err;
+
+static void get_fota_callback(int16_t result_code,
+		     size_t offset, const uint8_t *payload, size_t len,
+		     bool last_block, void *user)
+{
+	LOG_DBG("result_code: %d.%02d, offset:0x%X, len:0x%X, last_block:%d",
+		result_code / 32u, result_code & 0x1f, offset, len, last_block);
+	if (result_code != COAP_RESPONSE_CODE_CONTENT) {
+		fota_err = result_code;
+	} else if (payload && len) {
+		LOG_INF("Got FOTA response: %.*s", len, (const char *)payload);
+		fota_err = coap_codec_fota_resp_decode(user, payload, len,
+						       COAP_CONTENT_FORMAT_APP_JSON);
+	} else {
+		fota_err = 0;
```

Comment:
I think this can be removed.  It was handling the case that there was no FOTA job (COAP_RESPONSE_CODE_CONTENT) but also no payload (empty).  But the changes on the CoAP server now result in a 4.04 (NOT FOUND) result code when there is no FOTA job.